processing/src/main/java/org/apache/druid/query/groupby/epinephelinae/column/DictionaryBuildingGroupByColumnSelectorStrategy.java

```diff
@@ -0,0 +1,203 @@
+   * {@link #dictionary}.
+   * Absence of mapping of a "value" (denoted by returning {@link GroupByColumnSelectorStrategy#GROUP_BY_MISSING_VALUE})
+   * represents that the value is absent in the dictionary
+   */
+  private final Object2IntMap<DimensionType> reverseDictionary;
+
+  private DictionaryBuildingGroupByColumnSelectorStrategy(
+      DimensionToIdConverter<DimensionType> dimensionToIdConverter,
+      ColumnType columnType,
+      NullableTypeStrategy<DimensionType> nullableTypeStrategy,
+      DimensionType defaultValue,
+      IdToDimensionConverter<DimensionType> idToDimensionConverter,
+      List<DimensionType> dictionary,
+      Object2IntMap<DimensionType> reverseDictionary
+  )
+  {
+    super(dimensionToIdConverter, columnType, nullableTypeStrategy, defaultValue, idToDimensionConverter);
+    this.dictionary = dictionary;
+    this.reverseDictionary = reverseDictionary;
+  }
+
+  /**
+   * Creates an implementation of the strategy for the given type
+   */
+  public static GroupByColumnSelectorStrategy forType(final ColumnType columnType)
+  {
+    if (columnType.equals(ColumnType.STRING)) {
+      // String types are handled specially because they can have multi-value dimensions
+      throw DruidException.defensive("Should use special variant which handles multi-value dimensions");
+    } else if (
+      // Defensive check, primitives should be using a faster fixed-width strategy
+        columnType.equals(ColumnType.DOUBLE)
+        || columnType.equals(ColumnType.FLOAT)
+        || columnType.equals(ColumnType.LONG)) {
+      throw DruidException.defensive("Could used a fixed width strategy");
+    }
+
+    // Catch-all for all other types, that can only have single-valued dimensions
+    return forArrayAndComplexTypes(columnType);
+  }
+
+  /**
+   * Implemenatation of dictionary building strategy for types other than strings (since they can be multi-valued and need
+   * to be handled separately) and numeric primitives (since they can be handled by fixed-width strategy).
+   * This also means that we handle array and complex types here, which simplifies the generics a lot, as everything can be
+   * treated as Object in this class.
+   * <p>
+   * Also, there isn't any concept of multi-values here, therefore Dimension == DimensionHolderType == Object. We still
+   * homogenize rogue selectors which can return non-standard implementation of arrays (like Long[] for long arrays instead of
+   * Object[]) to what the callers would expect (i.e. Object[] in this case).
+   */
+  private static GroupByColumnSelectorStrategy forArrayAndComplexTypes(final ColumnType columnType)
+  {
+    final List<Object> dictionary = DictionaryBuildingUtils.createDictionary();
+    final Object2IntMap<Object> reverseDictionary =
+        DictionaryBuildingUtils.createReverseDictionary(columnType.getNullableStrategy());
+    return new DictionaryBuildingGroupByColumnSelectorStrategy<>(
```

Comment:
Also what would happen in Array<Strings> earlier we had an optimization where each string element of the array of dictionary encoded and the whole array was made of of Array<dictionaryInts> which gave us best compression. In your case we would have only one level of dictionary.

Reply:
Great catch! I have corrected the code to use a multi-level dictionary in the case of array<string>. I'd assume that the changed code might be better in low-cardinality use case, or when the memory is abundant (due to single lookup), while the earlier code would have performed better in high cardinality use case, and/or when the memory is constrained (due to less frequent spills). 

The benchmarks suggest that the new approach might be better, but the data was skewed (["1"], ["2"], ...) and the strings within the columns weren't reused. I haven't tested it with the real-world data, and with different configurations, which might be required to vet this change. 

Since the purpose of this patch is to enable grouping on complex columns, I have modified the code to be inline with the original code, and I can do the benchmarking for this change later.